Command: Support posix_spawn() on FreeBSD/OSX/GNU Linux

bdrewery opened the pull request:
Other platforms may support it as well but I have only tested on FreeBSD. For Linux we will need to detect a "new enough" glibc that contains the changes in https://sourceware.org/bugzilla/show_bug.cgi?id=10354#c12 specifically for the change noted in (2).

bdrewery commented:
FYI @alexcrichton, didn't assign you since you wrote the majority of the changes here.

alexcrichton commented:
Nice! For posterity I originally didn't land this because it broke our test suite, notably the behavior of `Command::new("nonexistent").status()` changed. Today that returns an error whereas with `posix_spawn` (on Linux at least) it return success and then `wait()` returns an error.

@bdrewery you were mentioning that a `try_wait` may be necessary here? Or does FreeBSD's implementation of `posix_spawn` already handle that case differently?

bdrewery commented:
FreeBSD's `posix_spawn` simply returns any error from `exec` already without any `wait` needed. The tests for this are passing:
```
test [run-pass] run-pass/command-before-exec.rs ... ok
test [run-pass] run-pass/command-exec.rs ... ok
```

It's able to do this because of `vfork` suspending the child and sharing memory, so it is able to share the errno:
```
static int
do_posix_spawn(pid_t *pid, const char *path,
    const posix_spawn_file_actions_t *fa,
    const posix_spawnattr_t *sa,
    char * const argv[], char * const envp[], int use_env_path)
{
        pid_t p;
        volatile int error = 0;

        p = vfork();
        switch (p) {
        case -1:
                return (errno);
        case 0:
                if (sa != NULL) {
                        error = process_spawnattr(*sa);
                        if (error)
                                _exit(127);
                }
                if (fa != NULL) {
                        error = process_file_actions(*fa);
                        if (error)
                                _exit(127);
                }
                if (use_env_path)
                        _execvpe(path, argv, envp != NULL ? envp : environ);
                else
                        _execve(path, argv, envp != NULL ? envp : environ);
                error = errno;
                _exit(127);
        default:
                if (error != 0)
                        _waitpid(p, NULL, WNOHANG);
                else if (pid != NULL)
                        *pid = p;
                return (error);
        }
}
```
This is intended and documented behavior.
```
     1.   If posix_spawn() and posix_spawnp() fail for any of the reasons that would cause vfork() or one of the exec to fail, an error value is returned as described by vfork() and exec, respectively (or, if the error occurs after the calling process successfully returns, the
          child process exits with exit status 127).
```

alexcrichton commented:
Ah ok! I think I must have misunderstood then, that sounds perfect!

@bors: r+

bors commented:
:pushpin: Commit 85b82f2 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 85b82f254e28f5e50b25d8e096af0f01e4441ef7 -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 200, this pull request will be tested once the tree is reopened

cuviper commented:
> For Linux we will need to detect a "new enough" glibc that contains the changes in https://sourceware.org/bugzilla/show_bug.cgi?id=10354#c12 specifically for the change noted in (2).

Just for reference, those commits are contained in glibc-2.24.

cuviper commented:
Yes, [`gnu_get_libc_version()`](http://man7.org/linux/man-pages/man3/gnu_get_libc_version.3.html).  Or the POSIX `confstr(_CS_GNU_LIBC_VERSION, buf, len)` might be easier for portability, returning `EINVAL` for that name elsewhere.

alexcrichton commented:
Oh nice catch @cuviper! I was testing locally with 2.23 when I initially wrote this, and I found that this program:

```rust
use std::process::Command;                                
                                                          
fn main() {                                               
    println!("{:?}", Command::new("nonexistent").spawn());
}                                                         
```

where on today's standard library it prints an error but that's with running glibc 2.23 locally. Taking the same binary and running it with glibc 2.24 it prints an error (as today's rustc does).

I haven't tested on OSX yet but sounds like we can certainly do this for Linux too! We'll just need to check at runtime that the version is >= 2.24.

I also just tested with musl and it looks like they've implemented it in such a way (presumably the same way?) that prints an error as well, so on musl we can unconditionally use posix_spawn.

alexcrichton commented:
@bdrewery would you be up for putting some of these checks in this PR? Or should I file a follow-up issue?

bdrewery commented:
@alexcrichton I had thought the command-exec.rs cases covered this but I see now they don't use `spawn` but plenty of others do but lack a nonexistent test. I'll add some more here. I think adding other platforms should be separate. I was considering looking at OSX and Linux but need to get some dev environments setup still.

alexcrichton commented:
@bdrewery oh I'm not too worried about test coverage here, when I tested this on Linux I originally got a lot of failures. If this gets past @bors I'd be pretty confident that it's working correctly!

(FWIW I tested OSX and I needed [this diff](https://gist.github.com/anonymous/ecf20b474c85f797d025094c1b9cfef9) to get errors out but it otherwise worked)

bdrewery commented:
It looks like `posix_spawn` always returns `errno` directly, even on FreeBSD. Will retest there with `from_raw_os_error`.

I was just looking over OSX's opensource code and their `posix_spawn` manpage and thought it would work fine given they document these possible errors:
```
     [ENOENT]           The new process file does not exist.

     [ENOEXEC]          The new process file has the appropriate access permission, but has an unrecognized format (e.g., an invalid magic number in its header).
```
It seems like they based their implementation on NetBSD's.
I don't expect this to be workable on OpenBSD since they use `fork` rather than `vfork`.
DragonFlyBSD appears to have the same implementation of FreeBSD so should work fine but still need to copy over the bindings.

alexcrichton commented:
@bdrewery also to be clear I don't want to drag you into a bunch of portability nightmare scenarios, if you'd like to just do FreeBSD and be done with it that's perfectly ok!

bdrewery commented:
@alexcrichton yup understood, I don't mind looking at some of these right now.

bdrewery commented:
Tests pass on OSX for me so I've added support for it in, and fixed the error handling. Working on Linux support still.

bdrewery commented:
Pretty sure I have the Linux/gnu support working. My VM with an older glibc properly avoids `posix_spawn`, but my newer VM is taking forever to build. I might just followup with a separate PR for Linux and other platforms if I get VMs for them setup.

One question I have for Linux/gnu is that `parse_glibc_version` in _src/libstd/sys/unix/net.rs_ already covers fetching the version. Where would be a good place to move this to be shared? I figured _src/libstd/sys_common_gnu_ but that's a misnamed libbacktrace helper.

alexcrichton commented:
Awesome!

That function may be best placed actually in `src/libstd/sys/unix/os.rs` and then it won't be reexported from `std::os` but it can be used internally on Unix

sanmai-NL commented on src/libstd/sys/unix/process/process_unix.rs:
```diff
@@ -237,7 +237,7 @@ impl Command {
 
     #[cfg(not(any(target_os = "linux", target_os = "macos")))]
-    fn posix_spawn(&mut self, stdio: &ChildPipes, envp: Option<&CStringArray>)
+    fn posix_spawn(&mut self, _stdio: &ChildPipes, _envp: Option<&CStringArray>)
```
Could you change this to `_`?

> Note that there is a subtle difference between using only _ and using a name that starts with an underscore. Something like _x still binds the value to the variable, whereas _ doesn’t bind at all. 

(https://doc.rust-lang.org/book/second-edition/ch18-03-pattern-syntax.html)

bdrewery replied:
Thanks, fixed.

bdrewery commented:
For the record I've confirmed outside of Rust that NetBSD and DragonFlyBSD `posix_spawn` behaves as we want and that OpenBSD does not using a simple .c test I had a friend run.
```
#include <assert.h>
#include <stdlib.h>
#include <errno.h>
#include <spawn.h>

extern char **environ;

int
main(void)
{
    pid_t pid;
    char *pargv[] = {"/nonexistent", NULL};
    int status = posix_spawn(&pid, "/nonexistent", NULL, NULL, pargv, environ);
    assert(status == ENOENT);
    return 0;
}
```

bdrewery commented:
@alexcrichton I've added in Linux GNU support now. The test passes on glibc 2.24 and 2.19. I'll let you handle musl as I don't have a test system for it. I've moved the NetBSD/DragonFlyBSD support to #48681. This should be ready to land whenever the tree is open.

alexcrichton commented:
@bors: r+

Awesome, thanks so much for your work here @bdrewery! I'd be more than happy to take care of musl after this. I think it'll just involve removing `"gnu"` [on these lines ](https://github.com/bdrewery/rust/blob/d740083fc8981ee933dc48a6b3dcee21b82c993e/src/libstd/sys/unix/process/process_unix.rs#L239-L249) :)

bors commented:
:pushpin: Commit d740083 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d740083fc8981ee933dc48a6b3dcee21b82c993e -->

bdrewery commented:
Other PR I've had seemed to get merged fairly quickly. Is this still queued for auto merge? I updated the libc submodule here but would like to update it independently and bump the version to bring in https://github.com/rust-lang/libc/commit/78f93220d70e796ba4a3ab8482d0f57be763e8c2. So I've been waiting for this merge.

alexcrichton commented:
@bdrewery yeah unfortunately our queue has been quite slow as of late, but no worries this is [still in the queue](https://buildbot2.rust-lang.org/homu/queue/rust)!

If you'd liek to throw that in here though I can r+

bdrewery commented:
@alexcrichton I'll just wait as I'll probably conflict on Cargo.lock.

alexcrichton commented:
@bors: r-

ah unfortunately I think this [breaks OSX](https://travis-ci.org/rust-lang/rust/jobs/350194065), specifically `posix_spawn` doesn't look at the child's `PATH` for spawning the resulting executable, whereas `fork`/`exec` apparently does?

cemeyer commented:
To pull just the relevant part from that giant log:

```
[01:50:10] failures:
[01:50:10]
[01:50:10] ---- [run-pass] run-pass-fulldeps/issue-15149.rs stdout ----
[01:50:10]
[01:50:10] error: test run failed!
[01:50:10] status: exit code: 101
[01:50:10] command: "/Users/travis/build/rust-lang/rust/build/i686-apple-darwin/test/run-pass-fulldeps/issue-15149.stage2-i686-apple-darwin"
[01:50:10] stdout:
[01:50:10] ------------------------------------------
[01:50:10]
[01:50:10] ------------------------------------------
[01:50:10] stderr:
[01:50:10] ------------------------------------------
[01:50:10] thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', libcore/result.rs:945:5
```

(No test code line number is included.)

bdrewery commented:
@alexcrichton It's unclear to me, did this pass on Linux?

cuviper commented:
> specifically `posix_spawn` doesn't look at the child's `PATH` for spawning the resulting executable, whereas `fork`/`exec` apparently does?

AFAICT on Linux, `posix_spawn`, `posix_spawnp`, and `execvpe` will use the caller's current `PATH`, not the `PATH` in the `envp` argument.  Rust currently uses `fork`+`execvp`, not the GNU `execvpe`, and instead overwrites `environ` beforehand - so when `execvp` is called, it's already using the new `PATH`.

So, I guess a new `PATH` should disqualify using `posix_spawn`, at least if `argv[0]` doesn't contain a `/`.

bdrewery commented:
Hm yes that test fails on Linux as well.

bdrewery commented:
@cuviper yep that makes sense.

bdrewery commented:
More details in #15149

bdrewery commented:
I don't know rust well enough yet to get this right.
```
if let Some(p) = envp.find("PATH") {
    return Ok(None)
}
```
What's the proper way here with `envp: Option<&CStringArray>`? `CStringArray` seems pretty limited.

alexcrichton commented:
@bdrewery ah actually looking in there may be somewhat difficult. Alternatively though we could maintain a boolean like [`saw_nul`](https://github.com/bdrewery/rust/blob/d740083fc8981ee933dc48a6b3dcee21b82c993e/src/libstd/sys/unix/process/process_common.rs#L54) where you could have something like `may_have_changed_path` which defaults to `false` on the `CommandEnv` struct. Whenever `set` is called with the key as `PATH` or `clear` is called, for example, we could set the flag and then if the flag is set then posix_spawn is skipped.

bdrewery commented:
@alexcrichton Thanks, I'll try that.

bors commented:
:umbrella: The latest upstream changes (presumably #48859) made this pull request unmergeable. Please resolve the merge conflicts.

shepmaster commented:
Ping from triage, @bdrewery ! Will you have time soon to address the merge conflicts and any remaining feedback?

bdrewery commented:
@shepmaster Yep, have just been waiting on some other merges to make testing on FreeBSD less of a hassle.

shepmaster commented:
> Yep, have just been waiting on some other merges

Please list those here; we can then track this issue as blocked on them.

bdrewery commented:
@shepmaster #42681 is the tracking bug for them and is enough to list. It's only a blocker in the sense that without it I have to jump through hoops to build/test on FreeBSD which is my main development platform.

bdrewery commented:
@alexcrichton I've updated with your suggestion to avoid `posix_spawn` if PATH is set.  The _issue-15149_ test passes now and I still get `posix_spawn` being used in my own tests as seen in syscall traces for `vfork`. I think @cmeyer's suggestion is good but I just can't figure out the right things in rust to do it at this moment since I'm still learning.
This adds a dependency on #42681 since testing was too painful without it.

alexcrichton commented on src/libstd/sys_common/process.rs:
```diff
@@ -121,4 +125,12 @@ impl<K: EnvKey> CommandEnv<K> {
+        self.saw_path || self.clear
+    }
+    fn maybe_saw_path(&mut self, key: &OsStr) {
+        if !self.saw_path && key.to_os_string() == OsString::from("PATH") {
```
I think here you should be able to do a direct comparison with `key == "PATH"`?

bdrewery replied:
Yeah you're right. Weird, I fought with this for far too long :).

bdrewery replied:
Fix pushed.

alexcrichton commented:
@bors: r+

Awesome, thanks!

bors commented:
:pushpin: Commit 8e0faf7 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 8e0faf79c0859f22d4e11268f272247a6ef73709 -->

kennytm commented:
@bors r-

The `env_saw_path` function is unused in musl environment, causing an unused item warning.

```
[00:40:43] error: method is never used: `env_saw_path`
[00:40:43]    --> libstd/sys/unix/process/process_common.rs:187:5
[00:40:43]     |
[00:40:43] 187 |     pub fn env_saw_path(&self) -> bool {
[00:40:43]     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:40:43]     |
[00:40:43] note: lint level defined here
[00:40:43]    --> libstd/lib.rs:232:31
[00:40:43]     |
[00:40:43] 232 | #![cfg_attr(not(stage0), deny(warnings))]
[00:40:43]     |                               ^^^^^^^^
[00:40:43]     = note: #[deny(dead_code)] implied by #[deny(warnings)]
```

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 70559c5 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 70559c54ce2a493a15f447436e281e16fc55b291 -->
